syntax = "proto3";

package game;

enum PacketID {
// Client → Server
    CS_Attack = 0;
    CS_ChangeWeapon = 1;
    CS_ItemPicked = 2;
    CS_KeyInput = 3;
    CS_PosInterpolation = 4;
    CS_RequestRestart = 5;
    CS_SendMessageAll = 6;
    CS_SendMessageTeam = 7;
    CS_SendNickname = 8;
    CS_ShotHit = 9;

// Server → Client
    SC_Attack = 10;
    SC_ChangeWeapon = 11;
    SC_CharacterDown = 12;
    SC_CharacterKillLog = 13;
    SC_CreateMyCharacter = 14;
    SC_CreateOtherCharacter = 15;
    SC_ItemPickFail = 16;
    SC_ItemPickSuccess = 17;
    SC_ItemSpawned = 18;
    SC_KeyInput = 19;
    SC_OnAccept = 20;
    SC_PosInterpolation = 21;
    SC_SendMessageAll = 22;
    SC_SendMessageTeam = 23;
    SC_ShotHit = 24;
}

message CS_ATTACK {
    uint32 hittedTargetId = 1; // 피격 대상
    float normalX = 2; // 이펙트의 노멀 벡터
    float normalY = 3;
    float normalZ = 4;
    float posX = 5; // 이펙트가 터질 위치
    float posY = 6;
    float posZ = 7;
}
message CS_CHANGE_WEAPON {
    uint32 weapon = 1; // 무기 교체
}
message CS_ITEM_PICKED {
    uint32 itemId = 1; // 획득 시도한 아이템의 정보
}
message CS_KEY_INPUT {
    bool keyW = 1; // 키(WASD)
    bool keyA = 2;
    bool keyS = 3;
    bool keyD = 4;
    float rotateAxisX = 5; // 마우스 X축 회전값
    float rotateAxisY = 6; // 마우스 Y축 회전값
    uint32 Jump = 7; // 점프 여부
    float normalX = 8; // 플레이어의 노멀 벡터
    float normalY = 9;
    float normalZ = 10;
}
message CS_POS_INTERPOLATION {
    float posX = 1; // 플레이어의 현재 위치
    float posY = 2;
    float posZ = 3;
}
message CS_REQUEST_RESTART {
    uint32 playerId = 1; // 플레이어의 고유 ID
    uint32 weapon = 2; // 들고 있는 무기 정보
}
message CS_SEND_MESSAGE_ALL {
    uint32 playerId = 1; // 플레이어의 고유 ID
    string message = 2; // 전송할 문자열
}
message CS_SEND_MESSAGE_TEAM {
    uint32 playerId = 1; // 플레이어의 고유 ID
    string message = 2; // 전송할 문자열
}
message CS_SEND_NICKNAME {
    string name = 1; // 최대 20자 유니코드 문자열 (실제 제한은 코드로 처리, 제한은 필요할때 늘리면 됨)
}
message CS_SHOT_HIT {
    uint32 playerId = 1; // 피격 대상
    uint32 hp = 2; // 피격 대상의 남은 HP
}
message KDAInfo {
    uint32 kill = 1;
    uint32 death = 2;
    uint32 assist = 3;
}
message PlayerInfo {
    uint32 playerId = 1;
    KDAInfo kda = 2;
}
message SC_ATTACK {
    uint32 playerId = 1; // CS_ATTACK을 전송한 플레이어 Id를 방에 있는 모든 플레이어들에게 전송
    uint32 hittedTargetId = 2; // 피격 대상
    float normalX = 3; // 이펙트의 노멀 벡터
    float normalY = 4;
    float normalZ = 5;
    float posX = 6; // 이펙트가 터질 위치
    float posY = 7;
    float posZ = 8;
}
message SC_CHANGE_WEAPON {
    uint32 playerId = 1; // 다른 캐릭터 ID
    uint32 weapon = 2; // 무기 교체
}
message SC_CHARACTER_DOWN {
    uint32 playerId = 1; // 서버로 부터 부여받은 ID
    uint32 teamID = 2; // 접속한 팀 ID
    repeated uint32 assistPlayerId = 3; // 어시스트 관련 
}
message SC_CHARACTER_KILL_LOG {
    repeated PlayerInfo playerInfoList = 1; // 모든 플레이어들의 KDA 정보
}
message SC_CREATE_MY_CHARACTER {
    uint32 playerId = 1; // 서버로 부터 부여받은 ID
    uint32 posIndex = 2; // 생성될 위치 인덱스 정보
    uint32 maxHP = 3; // 최대 Hp
    uint32 teamID = 4; // 접속한 팀 ID
}
message SC_CREATE_OTHER_CHARACTER {
    uint32 playerId = 1; // 서버로 부터 부여받은 ID
    uint32 posIndex = 2; // 생성될 위치 인덱스 정보
    uint32 maxHP = 3; // Hp
    uint32 curHP = 4;
    string name = 5; // 이름
    KDAInfo kdaInfo = 6; // KDA 정보
    uint32 weapon = 7; // 들고 있는 무기 정보
    uint32 teamID = 8; // 접속한 팀 ID
}
message SC_ITEM_PICK_FAIL {
    uint32 playerId = 1; // 획득 실패한 플레이어의 고유 ID
    uint32 itemId = 2; // 아이템의 고유 ID
}
message SC_ITEM_PICK_SUCCESS {
    uint32 playerId = 1; // 획득한 플레이어의 고유 ID
    uint32 itemId = 2; // 아이템의 고유 ID
}
message SC_ITEM_SPAWNED {
    uint32 itemId = 1; // 생성될 아이템의 고유 ID
    uint32 itemType = 2; // 생성될 아이템의 타입
    uint32 itemPosIndex = 3; // 아이템이 생성될 장소에 대한 정보
}
message SC_KEY_INPUT {
    uint32 playerId = 1;
    bool keyW = 2; // 키(WASD)
    bool keyA = 3;
    bool keyS = 4;
    bool keyD = 5;
    float rotateAxisX = 6; // 마우스 X축 회전값
    float rotateAxisY = 7; // 마우스 Y축 회전값
    uint32 Jump = 8; // 점프 여부
    float normalX = 9; // 플레이어의 노멀 벡터
    float normalY = 10;
    float normalZ = 11;
}
message SC_ON_ACCEPT {
    uint32 playerId = 1; // 서버로 부터 부여받은 ID
}
message SC_POS_INTERPOLATION {
    uint32 playerId = 1; // 플레이어 Id
    float posX = 2; // 플레이어의 현재 위치
    float posY = 3;
    float posZ = 4;
}
message SC_SEND_MESSAGE_ALL {
    uint32 playerId = 1; // 플레이어의 고유 ID
    string message = 2; // 전송할 문자열
}
message SC_SEND_MESSAGE_TEAM {
    uint32 playerId = 1; // 플레이어의 고유 ID
    string message = 2; // 전송할 문자열
}
message SC_SHOT_HIT {
    uint32 playerId = 1; // 피격 대상
    uint32 hp = 2; // 피격 대상의 남은 HP
}
